File: web/src/graphql/mutations/favouriteTransactionUpsert.ts
import {
  FavouriteTransactionUpsertMutation,
  FavouriteTransactionUpsertMutationVariables
} from "../../graphql-types"
import { gql } from "../../utils/gql"
import { MutationOptions, useMutation } from "../../utils/graphqlClient/useMutation"

const MUTATION = gql`
  mutation FavouriteTransactionUpsert(
    $name: String!
    $shop: String!
    $memo: String
    $priceCents: Int
    $accountId: ID
  ) {
    favouriteTransactionUpsert(
      input: {
        name: $name
        shop: $shop
        memo: $memo
        priceCents: $priceCents
        accountId: $accountId
      }
    ) {
      favouriteTransaction {
        id
        name
      }
    }
  }
`

export const useFavouriteTransactionUpsert = (
  options: MutationOptions<FavouriteTransactionUpsertMutation> = {}
) =>
  useMutation<FavouriteTransactionUpsertMutation, FavouriteTransactionUpsertMutationVariables>(
    MUTATION,
    options
  )

File: app/graphql/mutations/favourite_transaction_upsert.rb
# frozen_string_literal: true

module Mutations
  class FavouriteTransactionUpsert < BaseMutation
    authenticated

    description "Creates or updates a favourite transaction setting for a shop"

    field :favourite_transaction, Types::FavouriteTransactionType, null: false

    argument :name, String, required: true
    argument :shop, String, required: true
    argument :memo, String, required: false
    argument :price_cents, Integer, required: false
    argument :account_id, ID, required: false
    argument :category_id, ID, required: false

    def resolve(name:, shop:, memo: nil, price_cents: nil, account_id: nil, category_id: nil)
      record = current_user.favourite_transactions.new
      record.update!(name:, shop:, memo: memo || "", price_cents:, account_id:, category_id:)

      { favourite_transaction: record }
    end
  end
end


